extensions/2.0/Khronos/KHR_meshopt_compression/README.md

```diff
@@ -0,0 +1,675 @@
+
+The encoding of the third component allows to compute K for each vector independently from the bit representation, and must encode 1.0 precisely which is equivalent to `(1 << (K - 1)) - 1` as an integer; values of the third component that aren't equal to `(1 << (K - 1)) - 1` for a valid `K` are invalid and the result of decoding such vectors is unspecified.
+
+When storing a K-bit integer in a 8-bit of 16-bit component when K is not 8 or 16, the remaining bits (e.g. top 6 bits in case of K=10) must be equal to the sign bit; the valid range of the resulting integer is from `-max` to `max` where `max = (1 << (K - 1)) - 1`. The behavior of decoding values outside of that range is unspecified.
+
+The output of the filter is three decoded unit vector components, stored as 8-bit or 16-bit normalized integers, and the last input component verbatim.
+
+```
+void decode(intN_t input[4], intN_t output[4]) {
+	// input[2] encodes a K-bit representation of 1.0
+	float32_t one = input[2];
+
+	float32_t x = input[0] / one;
+	float32_t y = input[1] / one;
+	float32_t z = 1.0 - abs(x) - abs(y);
+
+	// octahedral fixup for negative hemisphere
+	float32_t t = min(z, 0.0);
+
+	x -= copysign(t, x);
+	y -= copysign(t, y);
+
+	// renormalize (x, y, z)
+	float32_t len = sqrt(x * x + y * y + z * z);
+
+	x /= len;
+	y /= len;
+	z /= len;
+
+	output[0] = round(x * INTN_MAX);
+	output[1] = round(y * INTN_MAX);
+	output[2] = round(z * INTN_MAX);
+	output[3] = input[3];
+}
+```
+
+`INTN_MAX` is equal to 127 when using 8-bit components (N is 8) and equal to 32767 when using 16-bit components (N is 16).
+
+`copysign` returns the value with the magnitude of the first argument and the sign of the second argument.
+
+`round` returns the nearest integer value, rounding halfway cases away from zero.
+
+## Filter 2: quaternion
+
+Quaternion filter allows to encode unit length quaternions using normalized 16-bit integers for all components, but allows control over the precision used for the components and provides better quality compared to naively encoding each component one by one.
+
+This filter is only valid if `byteStride` is 8.
+
+The input to the filter is three quaternion components, excluding the component with the largest magnitude, encoded as signed normalized K-bit integers (2 <= K <= 16, integers are stored in two's complement format), and an index of the largest component that is omitted in the encoding. The largest component is assumed to always be positive (which is possible due to quaternion double-cover). To allow per-element control over K, the last input element must explicitly encode 1.0 as a signed normalized K-bit integer, except for the least significant 2 bits that store the index of the maximum component.
```

Comment:
How is it possible to have K = 2 given that two least significant bits are used for the index?

Reply:
Thank you! The documented and asserted range in meshoptimizer encoder starts at K=4; K=2 is indeed impossible. With K=3, 1.0 encoded as a 3-bit signed integer is `0b011`; this technically is possible to decode (as any K-bit representation of 1.0 can be assumed to have lower 2 bits as 1), but results in a slightly awkward description because the encoded value may effectively not have a leading `1` bit... I believe this is why the minimum documented value is K=4, so I will update to that for now.

Reply:
Ignoring potential awkwardness of the description, would there be any benefit in supporting K = 3?

Reply:
K=3 has not been previously supported, yeah. I ran some numbers on errors using a random quaternion sequence:

<img width="1483" height="881" alt="test_results" src="https://github.com/user-attachments/assets/13e323dc-ac4a-4a4e-87fd-50dc83110428" />

And the same test using a constant angular velocity rotation around a fixed (non-cardinal) axis:

<img width="1483" height="881" alt="test_results_sequence" src="https://github.com/user-attachments/assets/0b6df3d9-1761-4e8c-8a56-dc90366587f8" />

In both of these the bit size is after filter & attribute compression. Since K=3 was not supported before, the error seems unusably large and, for constant velocity rotation, K=4 already works well wrt compressed size, it seems better to keep the K=4 cutoff.